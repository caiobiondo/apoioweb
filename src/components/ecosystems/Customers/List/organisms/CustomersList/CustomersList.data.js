import gql from 'graphql-tag';

const ITEMS_PER_PAGE = 10;

export const CustomersListQuery = gql`
  query CustomersListQuery($offset: Int!, $limit: Int!) {
    customers(offset: $offset, limit: $limit) {
      code
      name
      avatar
      email
      phone
      operator
    }
  }
`;
