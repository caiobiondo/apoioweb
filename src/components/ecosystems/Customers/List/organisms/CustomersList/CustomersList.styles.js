import styled from 'styled-components';
import { Responsive } from '@entria/components';
import { spSmall, spMedium, spPage } from 'styles/spacing';
import { gray300, white, orange100 } from 'styles/colors';

export const Wrapper = {
  flex: '1 1 0',
  flexDirection: 'column',
  justifyContent: 'center',
  overflowY: 'scroll',
};

export const cellStyle = {
  position: 'relative',
  padding: '0px 20px',
  lineHeight: '50px',
  color: gray300,
  fontSize: '15px',
  fontFamily: 'Roboto',
};

export const LinkStyle = {
  position: 'absolute',
  height: '100%',
  width: '100%',
  left: 0,
  top: 0,
};

export const TableWrapper = styled.div`
  padding: 20px 20px;

  table tr td:first-child div {
    opacity: 0;
    transition: all 0.2s ease;
  }

  tr {
    transition: all 0.2s ease;
    background: ${white};

    td:first-child .is-selected,
    td:first-child .is-selected * {
      opacity: 1;
    }
  }

  thead tr th {
    color: ${gray300} !important;
    font-size: 15px !important;
    font-weight: 400 !important;
  }

  tbody tr:hover {
    td:first-child div {
      opacity: 1;
    }

    background: #f6f6f6;
  }

  @media (max-width: ${Responsive.VIEWPORT.MEDIUM}px) {
    tr td,
    tr th {
      display: none;
    }

    tr th:first-child,
    tr td:first-child {
      display: table-cell;
      width: 1px !important;
    }

    tr th:nth-child(2),
    tr td:nth-child(2) {
      display: table-cell;
    }

    td:first-child,
    td:first-child * {
      opacity: 1 !important;
    }
  }
`;

export const CustomerName = styled.div`
  display: flex;
`;

export const Avatar = styled.figure`
  display: block;
  border-radius: 100%;
  text-align: center;
  width: 50px;
  min-width: 50px;
  height: 50px;
  line-height: 50px;
  font-size: 17px;
  color: ${white};
  background: ${orange100};
  margin-left: 0;
  overflow: hidden;
  margin-right: 20px;

  img {
    width: 100%;
    height: 100%;
  }
`;

export const NameLabel = styled.figure`
  display: block;
  height: 50px;
  line-height: 54px;
  margin-left: 0;
  margin-right: 0;
`;

export const List = styled.ul`
  padding-left: 0;
  padding-top: ${spMedium};
  margin: ${spMedium};

  @media (max-width: ${Responsive.VIEWPORT.MEDIUM}px) {
    margin: ${spSmall};
  }
`;

export const LoadingWrapper = styled.div`
  text-align: center;
`;

export const scrolledContainer = {
  flex: '1 1 0',
  flexDirection: 'column',
  justifyContent: 'center',
  paddingBottom: spPage,
};

export const fullContainer = {
  height: '100%',
};
