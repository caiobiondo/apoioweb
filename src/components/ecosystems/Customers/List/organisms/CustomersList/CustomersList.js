import React, { Component } from 'react';
import { Loading, CircularProgress, Paper, Table } from 'natura-ui';
import { graphql } from 'react-apollo';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router-dom';
import Checkbox from 'components/atoms/Checkbox';

import {
  CustomerName,
  LoadingWrapper,
  fullContainer,
  cellStyle,
  SelectedRowStyle,
  Wrapper,
  TableWrapper,
  NameLabel,
  Avatar,
  LinkStyle,
} from './CustomersList.styles';
import { CustomersListQuery } from './CustomersList.data';
import EmptyList from 'components/molecules/EmptyList/EmptyList';

export class CustomersList extends Component {
  loader = (
    <LoadingWrapper>
      <CircularProgress thickness={2} />
    </LoadingWrapper>
  );

  state = {
    data: {
      columns: ['select', 'name', 'email', 'phone', 'operator'],
      style: {
        select: {
          width: '1px',
          ...cellStyle,
        },
        name: {
          ...cellStyle,
          color: '#222',
          fontSize: '17px',
          fontWeight: '500',
        },
        email: cellStyle,
        phone: cellStyle,
        operator: cellStyle,
      },
      renderer: {
        name: this.renderName,
        email: this.renderCell,
        phone: this.renderCell,
        operator: this.renderCell,
        select: this.renderSelect,
        thead: {
          select: this.renderSelectAll,
        },
      },
      header: {
        name: <FormattedMessage id="customerName" />,
        email: <FormattedMessage id="customerEmail" />,
        phone: <FormattedMessage id="customerPhone" />,
        operator: <FormattedMessage id="customerPhoneProvider" />,
      },
    },
  };

  componentWillReceiveProps({ data, selectedCustomers }) {
    const { customers } = data;

    const parsedCustomers = (customers || []).map(customer => ({
      id: customer.id,
      name: customer.name,
      email: customer.emails && customer.emails.length && customer.emails[0].email,
      phone: customer.phones && customer.phones.length && customer.phones[0].phone,
      operator: customer.phones && customer.phones.length && customer.phones[0].provider,
      avatar: '',
      style: this.isSelected(customer, selectedCustomers) ? SelectedRowStyle : null,
    }));

    this.setState({ data: { ...this.state.data, body: parsedCustomers } });
  }

  isLoading(loading, orders) {
    return loading && !orders;
  }

  isEmpty(loading, orders) {
    return !loading && (!orders || orders.length === 0);
  }

  isSelected = (customer, selectedCustomers) => {
    return selectedCustomers.filter(({ id }) => id === customer.id).length;
  };

  selectCustomer = customer => {
    const { select } = this.props;
    select(customer);
  };

  selectAllCustomers = () => {
    const { select } = this.props;
    const { customers } = this.props.data;
    select(customers);
  };

  renderCell = ({ value, row }) => {
    return (
      <div>
        {value}
        <Link style={LinkStyle} to={`/my-customers/${row.id}`} />
      </div>
    );
  };

  renderName = ({ value, row }) => {
    const nameInitials = value
      .replace(/[^a-zA-Z- ]/g, '')
      .match(/\b\w/g)
      .join('');
    return (
      <CustomerName>
        <Avatar>{row.avatar ? <img src={row.avatar} alt={value} /> : nameInitials}</Avatar>
        <NameLabel>
          {value}
          <Link style={LinkStyle} to={`/my-customers/${row.id}`} />
        </NameLabel>
      </CustomerName>
    );
  };

  renderSelect = ({ row }) => {
    const isSelected = this.props.selectedCustomers.find(custmr => custmr.id === row.id);
    return (
      <div className={isSelected ? 'is-selected' : ''} onClick={() => this.selectCustomer(row)}>
        <Checkbox checked={isSelected} />
      </div>
    );
  };

  renderSelectAll = () => {
    return (
      <div onClick={() => this.selectAllCustomers()}>
        <Checkbox />
      </div>
    );
  };

  render() {
    const { isLoading, isEmpty } = this;
    const { loading, customers } = this.props.data;
    const { data } = this.state;

    if (isLoading(loading, customers)) {
      return <Loading background="transparent" />;
    }

    if (isEmpty(loading, customers)) {
      return (
        <Paper style={fullContainer}>
          <EmptyList icon="ico_box" titleId="customersEmptyList" />
        </Paper>
      );
    }

    return (
      <Paper style={Wrapper}>
        <TableWrapper>
          <Table data={data} />
        </TableWrapper>
      </Paper>
    );
  }
}

export default graphql(CustomersListQuery)(CustomersList);
