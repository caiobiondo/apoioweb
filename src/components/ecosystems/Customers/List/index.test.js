import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
import CustomersList from './index';

describe('CustomersList Ecosystem', () => {
  it('should render the customers list page', () => {
    const renderer = new ShallowRenderer();

    renderer.render(<CustomersList />);
    const result = renderer.getRenderOutput();

    expect(result).toMatchSnapshot();
  });

  it('should set selectedCustomers to an empty array', () => {
    const originalState = { selectedCustomers: [{ id: 123 }] };
    const expectedState = { selectedCustomers: [], empty: false, loading: true };

    const result = shallow(<CustomersList />);
    const instance = result.instance();
    instance.setState(originalState, () => {
      instance.onRemoveCustomer();

      expect(instance.state).toEqual(expectedState);
    });
  });

  it('should add customer to selected customers empty array', () => {
    const expectedState = { selectedCustomers: [{ id: 123 }], empty: false, loading: true };
    const customer = { id: 123 };

    const result = shallow(<CustomersList />);
    const instance = result.instance();
    instance.onSelectCustomer(customer);

    expect(instance.state).toEqual(expectedState);
  });

  it('should add customer to selected customers', () => {
    const originalState = { selectedCustomers: [{ id: 123 }] };
    const expectedState = {
      selectedCustomers: [{ id: 123 }, { id: 124 }],
      empty: false,
      loading: true,
    };
    const customer = { id: 124 };

    const result = shallow(<CustomersList />);
    const instance = result.instance();
    instance.setState(originalState, () => {
      instance.onSelectCustomer(customer);

      expect(instance.state).toEqual(expectedState);
    });
  });

  it('should add a list of customers to selected customers', () => {
    const originalState = { selectedCustomers: [] };
    const expectedState = {
      selectedCustomers: [{ id: 124 }, { id: 125 }],
      empty: false,
      loading: true,
    };
    const customers = [{ id: 124 }, { id: 125 }];

    const result = shallow(<CustomersList />);
    const instance = result.instance();
    instance.setState(originalState, () => {
      instance.onSelectCustomer(customers);

      expect(instance.state).toEqual(expectedState);
    });
  });

  it('should remove a list of customers from selected customers', () => {
    const originalState = { selectedCustomers: [{ id: 124 }, { id: 125 }] };
    const expectedState = { selectedCustomers: [], empty: false, loading: true };
    const customers = [];

    const result = shallow(<CustomersList />);
    const instance = result.instance();
    instance.setState(originalState, () => {
      instance.onSelectCustomer(customers);

      expect(instance.state).toEqual(expectedState);
    });
  });

  it('should remove a customer from selected customers', () => {
    const originalState = { selectedCustomers: [{ id: 124 }, { id: 125 }] };
    const expectedState = { selectedCustomers: [{ id: 125 }], empty: false, loading: true };
    const customer = { id: 124 };

    const result = shallow(<CustomersList />);
    const instance = result.instance();
    instance.setState(originalState, () => {
      instance.onSelectCustomer(customer);

      expect(instance.state).toEqual(expectedState);
    });
  });

  it('should remove all customers from selected customers', () => {
    const originalState = { selectedCustomers: [{ id: 125 }] };
    const expectedState = { selectedCustomers: [], empty: false, loading: true };
    const customer = [{ id: 125 }];

    const result = shallow(<CustomersList />);
    const instance = result.instance();
    instance.setState(originalState, () => {
      instance.onSelectCustomer(customer);

      expect(instance.state).toEqual(expectedState);
    });
  });
});
