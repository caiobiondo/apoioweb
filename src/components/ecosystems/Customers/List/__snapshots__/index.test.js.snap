// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomersList Ecosystem should render the customers list page 1`] = `
<indexstyles__Main
  empty={false}
  loading={true}
>
  <Apollo(CustomersList)
    onLoadFinished={[Function]}
    onSelect={[Function]}
    selectedCustomers={Array []}
  />
</indexstyles__Main>
`;
